crucible-jvm/src/Lang/Crucible/JVM/Translation/Class.hs

```diff
@@ -810,56 +809,51 @@ newInstanceInstr cls fieldIds = do
     addField (f,i) fs =
       App (InsertStringMapEntry knownRepr fs f i)
 
--- Field access is tricky
--- Fields are named as "C.f" where C is the static of the object that is being accessed.
--- However, that is not necessarily the name of the field if it was inherited from a
--- superclass. So we need to first look for C.f, but if we don't find it, we need to check
--- for fields named by the superclass of C.
-findField :: (KnownRepr TypeRepr a) => Expr JVM s (StringMapType JVMValueType) -> J.FieldId
-          -> (J.FieldId -> Expr JVM s JVMValueType -> JVMGenerator s ret (Expr JVM s a))
-          -> JVMGenerator s ret (Expr JVM s a)
-findField fields fieldId k = do
-  let currClassName = J.fieldIdClass fieldId
-  let str    = fieldIdText fieldId
-  let key    = App (StringLit (UnicodeLiteral str))
-  let mval   = App (LookupStringMapEntry knownRepr fields key)
-  caseMaybe mval knownRepr
-   MatchMaybe
-   { onJust  = \val -> k fieldId val
-   , onNothing = do
-       cls <- lookupClassGen currClassName
-       case (J.superClass cls) of
-         Nothing    -> reportError $ App $ StringLit (UnicodeLiteral ("getfield: field " <> str <> " not found"))
-         Just super -> findField fields (fieldId { J.fieldIdClass = super }) k
-   }
+-- | Given a 'J.FieldId' from a field get or set instruction, consult
+-- the JVM context to determine the class where the field was actually
+-- declared (which may be a superclass of the one specified in the
+-- input 'J.FieldId'), as specified in section 5.4.3.2 "Field
+-- Resolution" of the Java Virtual Machine Specification. Then return
+-- a key that can be used with an object's field map.
+resolveField :: J.FieldId -> JVMGenerator s ret (Expr JVM s (StringType Unicode))
+resolveField fieldId =
+  do cls <- lookupClassGen (J.fieldIdClass fieldId)
+     case any fieldMatch (J.classFields cls) of
+       True -> pure (App (StringLit (UnicodeLiteral (fieldIdText fieldId))))
```

Comment:
Just to make sure I understand: the result we compute here doesn't depend on which class the field is found in, correct?  In other words, do we need to search superclasses to compute the result here, or just to raise an error if the field doesn't exist?

Reply:
Yes, we do need to search through the superclasses; note that the class part of the `fieldId` changes in the recursive call. The point is to find the closest class in the chain of superclasses where a matching field exists.